Option to ignore existing pods' preferred inter-pod affinities if the incoming pod has no preferred inter-pod affinities

danielvegamyhre opened the pull request:
**What type of PR is this?**
/kind api-change
/sig scheduling
/sig api-machinery
/priority backlog

**What this PR does / why we need it:**
The purpose of these changes are to add an option for the scheduler to ignore existing pods` preferred inter-pod affinities if the incoming pod has no preferred inter-pod affinities. 

Preferred inter-pod affinity/anti-affinity rules tend to be made in both directions (i.e. if pod type A has a preferred affinity for pod type B, then pod type B will usually have a preferred affinity for pod type A). If an incoming pod that needs to be scheduled has no preferred inter-pod affinity rules, it is most likely the case that no existing pods have no preferred inter-pod affinity rules relating to it. 

Therefore, the linear scan through existing pods checking their preferred inter-pod affinities is usually a waste of time that the user can choose to optimize away by enabling this option in the InterPodAffinity plugin config args (at the cost of an occasional pod being scheduled non-optimally/violating existing pods affinity rules). 

Specifically, this option will cause [PreScore](https://github.com/kubernetes/kubernetes/blob/3e26e104bdf9d0dc3c4046d6350b93557c67f3f4/pkg/scheduler/framework/plugins/interpodaffinity/scoring.go#L126) to exit early if this option is enabled and the incoming pod has no existing inter-pod affinity rules. 

**Which issue(s) this PR fixes:**
Fixes #114267 

**Does this PR introduce a user-facing change?**
```release-note
Added new option to the InterPodAffinity scheduler plugin to ignore existing pods` preferred inter-pod affinities if the incoming pod has no preferred inter-pod affinities. This option can be used as an optimization for higher scheduling throughput (at the cost of an occasional pod being scheduled non-optimally/violating existing pods' preferred inter-pod affinities). To enable this scheduler option, set the InterPodAffinity scheduler plugin arg "ignorePreferredTermsOfExistingPods: true".
```

danielvegamyhre committed:
optionally ignore preferred terms of existing pods unless incoming pod
has inter-pod affinities

ahg-g commented:
/assign
/ok-to-test

danielvegamyhre commented:
@ahg-g could you take a quick look at this when you have a moment please?

ahg-g commented on pkg/scheduler/apis/config/types.go:
```diff
@@ -87,6 +87,10 @@ type KubeSchedulerConfiguration struct {
 
+	// IgnoreExistingInterPodAffinities configures the scheduler to ignore existing pods' affinity
+	// rules when scoring candidate nodes, unless the incoming pod has inter-pod affinities.
+	IgnoreExistingInterPodAffinities bool
```
Remove this field pls, we only need the one under InterPodAffinityArgs

danielvegamyhre replied:
Done

ahg-g commented on pkg/scheduler/apis/config/types_pluginargs.go:
```diff
@@ -52,6 +52,10 @@ type InterPodAffinityArgs struct {
+
+	// IgnoreExistingInterPodAffinities configures the scheduler to ignore existing pods' affinity
+	// rules when scoring candidate nodes, unless the incoming pod has inter-pod affinities.
+	IgnoreExistingInterPodAffinities bool
```
The name should indicate that this is only related to preferred terms; so perhaps we can name it `IgnorePreferredTermsOfExistingPods`

danielvegamyhre replied:
Makes sense - done.

ahg-g commented on pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go:
```diff
@@ -929,3 +929,120 @@ func TestPreferredAffinityWithHardPodAffinitySymmetricWeight(t *testing.T) {
+			name: "incoming pod does not have inter-pod affinities",
+			pod:  &v1.Pod{ObjectMeta: metav1.ObjectMeta{Labels: podLabelServiceS1}},
+			pods: []*v1.Pod{
+				{Spec: v1.PodSpec{NodeName: "node1"}},
```
pls make the two tests exactly the same with the only difference being the value of `IgnoreExistingInterPodAffinities`

ahg-g replied:
Also, there is no need to add a new test, just add the cases to the existing one: https://github.com/kubernetes/kubernetes/blob/e74d62ed3010c9244e55a653fecc89122a208ced/pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go#L371

danielvegamyhre replied:
Ok I removed the new test and added two test cases to the existing `TestPreferredAffinity`, which are the same except for the value of `IgnorePreferredTermsOfExistingPods`: 

- `IgnorePreferredTermsOfExistingPods: true` which verifies node scores are all 0
- `IgnorePreferredTermsOfExistingPods: false` which verifies the correct node receives the highest score

danielvegamyhre commented:
Thanks for the review @ahg-g I have addressed all your comments, this is ready for another look. The only outstanding issue is in the test case I added to `options_test.go`. For some reason, despite the test case YAML setting`ignorePreferredTermsOfExistingPods: true`, the resulting scheduler config has it set as `false`. Any idea how this might be possible? 

I have tried both `bool` and `*bool` data types (based on other tests I saw using `*bool), as well as read through the YAML -> KubeSchedulerConfiguration code looking for where this might be occurring, but no luck so far.

ahg-g commented:
You need to update the auto-generated code that does the translation from the external to the internal types. Run `./hack/update-codegen.sh` to do that.

ahg-g commented on pkg/scheduler/apis/config/scheme/scheme_test.go:
```diff
@@ -1186,6 +1186,7 @@ profiles:
   - args:
       apiVersion: kubescheduler.config.k8s.io/v1beta2
       hardPodAffinityWeight: 5
+      ignorePreferredTermsOfExistingPods: false
```
worth adding a test to the scheme tests that sets it to true.

danielvegamyhre replied:
Ok I added test cases for both scheme encoding/decoding tests when `ignorePreferredTermsOfExistingPods` is true.

ahg-g commented on pkg/scheduler/framework/plugins/interpodaffinity/scoring.go:
```diff
@@ -142,6 +142,15 @@ func (pl *InterPodAffinity) PreScore(
+		cycleState.Write(preScoreStateKey, &preScoreState{
+			topologyScore: make(map[string]map[string]int64),
+		})
+		return nil
```
@alculquicondor @Huang-Wei are we going to allow returning `Skip` for PreScore extensions to indicate that Score shouldn't run for the same plugin?

Huang-Wei replied:
that sounds like a better contract, compared to plumbing the placeholder and checking if it's empty in Score.

alculquicondor replied:
yes, we should have support for Skip in Score as well, but the work didn't start yet.

ahg-g commented:
/retitle "Option to ignore existing pods' inter-pod affinities if the incoming pod has no inter-pod affinities"

ahg-g commented:
Please add a release note

alculquicondor commented:
Mention in the title, description and release note, that this refers to **preferred** pod affinity

danielvegamyhre commented:
I've updated the title, description, and release note to specify **preferred* inter-pod affinities.

ahg-g commented:
/label api-review

ahg-g commented:
looks good to me, lets wait for the api review now.

ahg-g commented:
/triage accepted
/priority backlog

cici37 commented:
/cc
Curious about the community requirement on adding field directly into stable API   :) 
Is that allowed because scheduler API is internal?

ahg-g commented:
> Curious about the community requirement on adding field directly into stable API :) Is that allowed because scheduler API is internal?

Adding field is allowed, removing them requires a new version.

alculquicondor commented:
Maybe the question is why no KEP? Since the component config API is not stored, we don't need feature gates or other things that a KEP imposes.

btw

/remove-sig api-machinery

this has nothing to do.

liggitt commented on pkg/scheduler/apis/config/types_pluginargs.go:
```diff
@@ -52,6 +52,10 @@ type InterPodAffinityArgs struct {
+
+	// IgnorePreferredTermsOfExistingPods configures the scheduler to ignore existing pods' preferred affinity
+	// rules when scoring candidate nodes, unless the incoming pod has inter-pod affinities.
+	IgnorePreferredTermsOfExistingPods bool
```
the "unless incoming pod..." aspect is not captured in the name... would we eventually want to also allow unconditionally ignoring preferred terms of existing pods? if so, it would be worth thinking about how we would name this field and that eventual field in ways that wouldn't be confusing

danielvegamyhre replied:
Based on my understanding, unconditionally ignoring preferred terms of existing pods is not something we'll want to do in the future. However, @ahg-g has more knowledge about future plans in this area so I will defer to his recommendation here.

ahg-g replied:
Right, I don't think we will want to do that

ahg-g commented:
Please squash the commits to one

ahg-g commented:
/lgtm

Thanks @danielvegamyhre
